azurerm/data_source_kubernetes_cluster.go

```diff
@@ -327,29 +363,109 @@ func dataSourceArmKubernetesClusterRead(d *schema.ResourceData, meta interface{}
 	return nil
 }
 
-func flattenKubernetesClusterDataSourceLinuxProfile(input *containerservice.LinuxProfile) []interface{} {
+func flattenKubernetesClusterDataSourceRoleBasedAccessControl(input *containerservice.ManagedClusterAADProfile) []interface{} {
+	if input == nil {
+		return []interface{}{}
+	}
+
+	profile := make(map[string]interface{})
+
+	if input.ClientAppID != nil {
+		profile["client_app_id"] = *input.ClientAppID
+	}
+
+	if input.ServerAppID != nil {
+		profile["server_app_id"] = *input.ServerAppID
+	}
+
+	if input.TenantID != nil {
+		profile["tenant_id"] = *input.TenantID
+	}
+
+	return []interface{}{
+		map[string]interface{}{
+			"azure_active_directory": []interface{}{
+				profile,
+			},
+		},
+	}
+}
+
+func flattenKubernetesClusterDataSourceAccessProfile(profile containerservice.ManagedClusterAccessProfile) (*string, []interface{}) {
+	if profile.AccessProfile == nil {
+		return nil, []interface{}{}
+	}
+
+	if kubeConfigRaw := profile.AccessProfile.KubeConfig; kubeConfigRaw != nil {
+		rawConfig := string(*kubeConfigRaw)
+		var flattenedKubeConfig []interface{}
+
+		if strings.Contains(rawConfig, "apiserver-id:") {
+			kubeConfigAAD, err := kubernetes.ParseKubeConfigAAD(rawConfig)
+
+			if err != nil {
+				return utils.String(rawConfig), []interface{}{}
+			}
+
+			flattenedKubeConfig = flattenKubernetesClusterDataSourceKubeConfigAAD(*kubeConfigAAD)
+		} else {
+			kubeConfig, err := kubernetes.ParseKubeConfig(rawConfig)
+
+			if err != nil {
+				return utils.String(rawConfig), []interface{}{}
+			}
+
+			flattenedKubeConfig = flattenKubernetesClusterDataSourceKubeConfig(*kubeConfig)
+		}
+
+		return utils.String(rawConfig), flattenedKubeConfig
+	}
+
+	return nil, []interface{}{}
+}
+
+func flattenKubernetesClusterDataSourceAddonProfiles(profile map[string]*containerservice.ManagedClusterAddonProfile) interface{} {
 	values := make(map[string]interface{})
-	sshKeys := make([]interface{}, 0)
 
-	if profile := input; profile != nil {
-		if username := profile.AdminUsername; username != nil {
-			values["admin_username"] = *username
+	routes := make([]interface{}, 0)
+	if httpApplicationRouting := profile["httpApplicationRouting"]; httpApplicationRouting != nil {
+		enabled := false
+		if enabledVal := httpApplicationRouting.Enabled; enabledVal != nil {
+			enabled = *enabledVal
 		}
 
-		if ssh := profile.SSH; ssh != nil {
-			if keys := ssh.PublicKeys; keys != nil {
-				for _, sshKey := range *keys {
-					if keyData := sshKey.KeyData; keyData != nil {
-						outputs := make(map[string]interface{})
-						outputs["key_data"] = *keyData
-						sshKeys = append(sshKeys, outputs)
-					}
-				}
-			}
+		zoneName := ""
+		if v := httpApplicationRouting.Config["HTTPApplicationRoutingZoneName"]; v != nil {
+			zoneName = *v
+		}
+
+		output := map[string]interface{}{
```

Comment:
This might be simpler as:

```go
    output := map[string]interface{}{
        "enabled":                            true,
        "http_application_routing_zone_name": "",
    }

    if v := httpApplicationRouting.Config["HTTPApplicationRoutingZoneName"]; v != nil {
        output["http_application_routing_zone_name"] = *v
    }

    if enabledVal := httpApplicationRouting.Enabled; enabledVal != nil {
        output["enabled"] = *v
    }
```

Reply:
it could - but I'd argue it's worth not redefining the keys here so I think we should leave this as-is for the moment